web-common/src/features/dashboards/stores/filter-utils.ts

```diff
@@ -0,0 +1,186 @@
+    },
+  } as V1Expression;
+}
+
+export function createBetweenExpression(
+  ident: string,
+  val1: number,
+  val2: number,
+  negate: boolean,
+) {
+  const exprs = [
+    {
+      cond: {
+        op: negate ? V1Operation.OPERATION_LTE : V1Operation.OPERATION_GT,
+        exprs: [{ ident }, { val: val1 }],
+      },
+    },
+    {
+      cond: {
+        op: negate ? V1Operation.OPERATION_GTE : V1Operation.OPERATION_LT,
+        exprs: [{ ident }, { val: val2 }],
+      },
+    },
+  ] as Array<V1Expression>;
+  if (negate) {
+    return createOrExpression(exprs);
+  } else {
+    return createAndExpression(exprs);
+  }
+}
+
+const conditionOperationComplement: Partial<Record<V1Operation, V1Operation>> =
+  {
+    [V1Operation.OPERATION_EQ]: V1Operation.OPERATION_NEQ,
+    [V1Operation.OPERATION_LT]: V1Operation.OPERATION_GTE,
+    [V1Operation.OPERATION_LTE]: V1Operation.OPERATION_GT,
+    [V1Operation.OPERATION_IN]: V1Operation.OPERATION_NIN,
+    [V1Operation.OPERATION_LIKE]: V1Operation.OPERATION_NLIKE,
+    [V1Operation.OPERATION_AND]: V1Operation.OPERATION_OR,
+  };
+// add inverse of existing values above
+for (const c in conditionOperationComplement) {
+  conditionOperationComplement[conditionOperationComplement[c]] = c;
+}
+
+export function negateExpression(expr: V1Expression): V1Expression {
+  if ("ident" in expr || "val" in expr || !expr.cond) return expr;
+  return {
+    cond: {
+      op:
+        conditionOperationComplement[expr.cond.op as V1Operation] ??
+        V1Operation.OPERATION_EQ,
+      exprs: expr.cond.exprs,
+    },
+  };
+}
+
+export function forEachExpression(
+  expr: V1Expression,
+  cb: (e: V1Expression, depth?: number) => void,
+  depth = 0,
+) {
+  if (!expr.cond?.exprs) {
+    cb(expr, depth);
+    return;
+  }
+
+  for (const subExpr of expr.cond.exprs) {
+    cb(subExpr, depth);
+    forEachExpression(subExpr, cb, depth + 1);
+  }
+}
+
+/**
+ * Creates a copy of the expression with sub expressions filtered based on {@link checker}
+ */
+export function filterExpressions(
+  expr: V1Expression,
+  checker: (e: V1Expression) => boolean,
+): V1Expression | undefined {
+  if (!expr.cond?.exprs) {
+    return {
+      ...expr,
+    };
+  }
+
+  const newExpr: V1Expression = {
+    cond: {
+      op: expr.cond.op,
+      exprs: expr.cond.exprs
+        .map((e) => filterExpressions(e, checker))
+        .filter((e) => e !== undefined && checker(e)) as Array<V1Expression>,
+    },
+  };
+
+  switch (expr.cond.op) {
+    // and/or will have only sub expressions
+    case V1Operation.OPERATION_AND:
+    case V1Operation.OPERATION_OR:
+      if (newExpr.cond?.exprs?.length === 0) return undefined;
+      break;
+
+    // other types will have identifier as 1st expression
+    default:
+      if (!newExpr.cond?.exprs?.length || !("ident" in newExpr.cond.exprs[0]))
+        return undefined;
+      break;
+  }
+
+  return newExpr;
+}
+
+export function getValueIndexInExpression(expr: V1Expression, value: any) {
```

Comment:
Hmmm i recently fixed an issue with non string dimensions. We should update the type everywhere and keep this as any for now.

Reply:
I really don't think this should be any. If it's not `string`, then I think it must be something like `string | number`. @begelundmuller -- can you weigh in here? If we've recently encountered error with non-string dimensions, can you help us out by telling us the complete universe of possible dimensions we might encounter?

FWIW, part of the reason I'm pushing on this is because the work on the downstream component code, for example all the stuff I did in the leaderboards last month, is made way more complicated if I can't have guarantees about the shape of the data used by the UI. It's a nightmare if we have to guard against everything that close to the display layer, and we should narrow these types closer to the data layer.